build: namespace generated headers with `zephyr/`

ycsin opened the pull request:
Namespaced the generated headers with `zephyr/` to prevent potential conflict with other headers.

Introduced a Kconfig (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH`) to support legacy generated header include paths in the mean time.

[Documentation preview](https://builds.zephyrproject.io/zephyr/pr/63973/docs/releases/migration-guide-3.7.html#build-system)

Fixes #60377
Fixes #68035

- [Architecture WG](https://github.com/zephyrproject-rtos/zephyr/pull/63973#issuecomment-1799139397)
- [x] Missing changes to the N repos that depend on this are required in order to merge the PR
   - [**sof**](https://github.com/thesofproject/sof)
     - https://github.com/thesofproject/sof/pull/8459
     - https://github.com/zephyrproject-rtos/sof/pull/34
   - [**percepio**](https://github.com/percepio/percepio)
     - https://github.com/percepio/percepio/pull/5
     - https://github.com/percepio/percepio/pull/6
     - https://github.com/zephyrproject-rtos/percepio/pull/10
     - https://github.com/percepio/percepio/pull/9
     - https://github.com/percepio/percepio/pull/8
     - https://github.com/zephyrproject-rtos/percepio/pull/13
   - https://github.com/zephyrproject-rtos/hal_stm32/pull/211
   - https://github.com/zephyrproject-rtos/hal_nxp/pull/395
   - https://github.com/zephyrproject-rtos/lvgl/pull/53
   - [**mcuboot**](https://github.com/mcu-tools/mcuboot)
     - https://github.com/mcu-tools/mcuboot/pull/1965
     - https://github.com/zephyrproject-rtos/mcuboot/pull/117

Automation script:

```python
#!/usr/bin/env python

from pathlib import Path
import re

EXTENSIONS = ("c", "h", "cpp", "rst", "md", "S", "ld", "py", "svg")

HEADERS = [
    "app_version",
    # "autoconf",
    "cmake_intdef",
    "core-isa-dM",
    "devicetree_generated",
    "driver-validation",
    "kobj-types-enum",
    "linker-kobject-prebuilt-data",
    "linker-kobject-prebuilt-priv-stacks",
    "linker-kobject-prebuilt-rodata",
    "mcuboot_version",
    "offsets",
    "otype-to-size",
    "otype-to-str",
    "strerror_table",
    "strsignal_table",
    "syscall_list",
    "version",
    "zsr",
]

HEADERS_H = [hdr + ".h" for hdr in HEADERS]

for p in Path(".").glob("zephyr/**/*"):
    if not p.is_file() or p.suffix and p.suffix[1:] not in EXTENSIONS:
        continue

    # skip git & github folders
    if (str(p).startswith("zephyr/.git")):
        continue

    # skip release docs
    if (str(p).startswith("zephyr/doc/releases")):
        continue

    content = ""
    with open(p, 'r', newline='', errors="surrogateescape") as f:
        for line in f:
            # for lines in sources
            m = re.match(r"^(.*)#include <(.*)>(.*)$", line)
            # for lines in docs
            n = re.match(r"^(.*)include/generated/(.*)\.h(.*)$", line)
            if (m and (m.group(2).startswith("syscalls/") or m.group(2) in HEADERS_H)):
                content += (m.group(1) + "#include <zephyr/" +
                            m.group(2) + ">" + m.group(3) + "\n")
            elif (n and n.group(2) in HEADERS):
                content += (n.group(1) + "include/generated/zephyr/" +
                            n.group(2) + ".h" + n.group(3) + "\n")
            else:
                content += line

    with open(p, "w", newline='') as f:
        f.write(content)

```

## [Requirements for Treewide Changes](https://docs.zephyrproject.org/latest/contribute/guidelines.html#requirements-for-treewide-changes)

- [x] The zephyr repository must apply the ‘treewide’ GitHub label to any issues or pull requests that are treewide changes
- [x] The person proposing a treewide change must create an [RFC issue](https://github.com/zephyrproject-rtos/zephyr/issues/new?assignees=&labels=RFC&template=003_rfc-proposal.md&title=) describing the change, its rationale and impact, etc. before any pull requests related to the change can be merged
- [x] The project’s [Architecture Working Group (WG)](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group) must include the issue on the agenda and discuss whether the project will accept or reject the change before any pull requests related to the change can be merged (with escalation to the TSC if consensus is not reached at the WG)
- [ ] The Architecture WG must specify the procedure for merging any PRs associated with each individual treewide change, including any required approvals for pull requests affecting specific subsystems or extra review time requirements
- [x] The person proposing a treewide change must email [devel@lists.zephyrproject.org](mailto:devel%40lists.zephyrproject.org) about the RFC if it is accepted by the Architecture WG before any pull requests related to the change can be merged

gmarull commented on CMakeLists.txt:
```diff
@@ -552,17 +552,25 @@ if(ZEPHYR_GIT_INDEX)
 add_custom_command(
-  OUTPUT ${PROJECT_BINARY_DIR}/include/generated/version.h
+  OUTPUT ${PROJECT_BINARY_DIR}/include/generated/zephyr/version.h
   COMMAND ${CMAKE_COMMAND} -DZEPHYR_BASE=${ZEPHYR_BASE}
-    -DOUT_FILE=${PROJECT_BINARY_DIR}/include/generated/version.h
+    -DOUT_FILE=${PROJECT_BINARY_DIR}/include/generated/zephyr/version.h
```
we should probably move all generated headers to `build/include/zephyr`, and include `build/include` (also `build/include/zephyr` to not break immediately).

gmarull replied:
it would be guarded by a Kconfig option, enabled by default now, deprecated later maybe (same as we did for the public includes a while back)

ycsin replied:
@henrikbrixandersen I think the `default y` came from here

henrikbrixandersen replied:
I see

ycsin commented:
this PR is getting more involved than I initially expected

fabiobaltieri commented:
@ycsin hey thanks for putting this together, since it's a major change I added the treewide and architecture review tags so this can be discussed in the architecture WG (cc @carlescufi)

de-nordic reviewed (dismissed):
Flash and MCUmgr look ok.

gmarull commented on Kconfig.zephyr:
```diff
@@ -911,3 +911,15 @@ config COMPAT_INCLUDES
+config LEGACY_GENERATED_INCLUDE_PATH
+	bool "Allow for the legacy include path for generated headers (without the zephyr/ prefix)"
+	default y
+	select DEPRECATED
```
I'd avoid `select DEPRECATED` in a first iteration, some builds enable `WARN_DEPRECATED`.

ycsin commented:
Rebased on top of `main`
Moved `LEGACY_GENERATED_INCLUDE_PATH` into the Compatibility menu

de-nordic reviewed (dismissed):
OK for Flash API and MCUmgr.

gmarull commented on CMakeLists.txt:
```diff
@@ -109,6 +109,10 @@ add_library(zephyr_interface INTERFACE)
 zephyr_library_named(zephyr)
 
+if(CONFIG_LEGACY_GENERATED_INCLUDE_PATH)
+  zephyr_include_directories(${PROJECT_BINARY_DIR}/include/generated/zephyr)
```
I'm fine with the PR, just one note for discussion (Arch WG): should generated headers have their own subfolder, ie, `<zephyr/generated/xxx.h>`? (maybe with some exceptions like `version.h`).

ycsin replied:
+1, had the same thing in my mind

gmarull replied:
Some will always be generated (e.g. syscall stuff, devicetree, etc.). I think it makes sense to put them into `generated/`, others like `version.h`, probably not.

cfriedt replied:
I guess the main concern is that a generated header could conflict with a static header, and that a generated header might incorrectly get included first.

Technically, that's still a risk today, so could be solved separately.

ycsin commented:
Another compilation issue getting fixed @ #64251

tejlmand reviewed (dismissed):
Thanks for this PR, i'm very much in favor of namespacing Zephyr headers.

That said, I see no justification for the changes to Kconfiglib.py, as kconfiglib.py is a project which originates outside of Zephyr.
Thus it should only be patched in Zephyr for good reasons, such as bug fixing or important feature extensions which cannot be done in other ways.

ycsin commented on scripts/kconfig/kconfiglib.py:
```diff
@@ -2928,7 +2928,7 @@ def _parse_block(self, end_token, parent, prev):
                 node = MenuNode()
                 node.kconfig = self
                 node.item = sym
-                node.is_menuconfig = (t0 is _T_MENUCONFIG)
```
Or should we leave this file as-is instead and merge this with superpower? cc @nashif

tejlmand replied:
after you reverted the changes pointed out in this comment https://github.com/zephyrproject-rtos/zephyr/pull/63973#discussion_r1368619501

then it should be as simple as dropping this commit https://github.com/zephyrproject-rtos/zephyr/pull/63973/commits/6ca708fb1d735482a73212de6c1f819a0145ba66 because you are no longer modifying kconfiglib.py.
So therefore CI should still be green (not considering other potential failures) 😉

mbolivar-ampere reviewed (dismissed):
Devicetree changes look great. Thank you for taking the time to update the documentation.

This +1 is limited to the DT changes.

fabiobaltieri commented:
Ran this again the push CI job in the testing repo (https://github.com/zephyrproject-rtos/zephyr-testing/runs/18077860182) single failure but it's unrelated. Great job there, hopefully we can merge it soon, too bad we did not make to this at the last arch wg, we should probably do it early next time, I don't see any objections here anyway.

Needs a rebase btw.

ycsin commented:
rebased and remove modifications to `scripts/kconfig/kconfiglib.py`

de-nordic reviewed (dismissed):
Flash API and MCUmgr look ok

ycsin commented:
rebased again for #64464

tejlmand reviewed (dismissed):
Approved, but please let the reviewer decide if review comments has been addressed (unless the reviewer has been unresponsive for an unreasonable long period).
![image](https://github.com/zephyrproject-rtos/zephyr/assets/9802583/8d9349ef-2d3b-44dd-bb72-45ab55edad43)

It was less than 24h since the last comment from me was addressed here:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/9802583/a711062e-f0e3-4131-8421-8654f57be75b)

carlescufi commented:
Architecture WG:

- Checked that we do not have an applicable Kconfig option anymore that can be used, instead we will need the new `CONFIG_LEGACY_GENERATED_INCLUDE_PATH`. 
- Missing changes to the 2 repos that depend on this are required in order to merge the PR
- Missing a entry to the migration document

cfriedt commented:
> this PR is getting more involved than I initially expected
> 
> EDIT: And CI is complaining about some lines that are not modified in this PR:

@ycsin - those compliance issues could also be fixed in a separate PR. This one will require a respin anyway, so that might be a good option

carlescufi commented:
> I'm not sure which category should this change falls into in the migration guide, any idea?

build system

ycsin committed:
doc: releases: add notes for namespaced generated headers

Add migration notes for generated headers being namespaced
to the `zephyr/` folder.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

henrikbrixandersen commented:
This needs a rebase. Is the target milestone still v3.6.0?

ycsin commented:
> This needs a rebase. Is the target milestone still v3.6.0?

I'm not sure tbh, does this need to go thru an architecture review?

henrikbrixandersen commented:
> I'm not sure tbh, does this need to go thru an architecture review?

Yes, it needs to follow the process described here: https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes

ycsin committed:
build: namespace the generated headers with `zephyr/`

Namespaced the generated headers with `zephyr` to prevent
potential conflict with other headers.

Introduce a temporary Kconfig `LEGACY_GENERATED_INCLUDE_PATH`
that is enabled by default. This allows the developers to
continue the use of the old include paths for the time being
until it is deprecated and eventually removed. The Kconfig will
generate a build-time warning message, similar to the
`CONFIG_TIMER_RANDOM_GENERATOR`.

Updated the includes path of in-tree sources accordingly.

Most of the changes here are scripted, check the PR for more
info.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

teburd commented:
Please do continue the work here, +1 from on CI and arch wg agreement

ycsin commented:
This PR should be ready for the architecture review now

ycsin committed:
build: namespace `autoconf.h` with `zephyr/`

Namespace the generated `autoconf.h` file with `zephyr/`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
testing the changes for `autoconf.h`, list of files that includes the `autoconf.h`:

```
(.venv) ycsin@LAPTOP-ROG:~/zephyrproject$ find . -type f -exec grep -H '#include [<|"]autoconf\.h[>|"]' {} +
./modules/hal/stm32/scripts/tests/genllheaders/data/stm32_ll_tim.h:#include <autoconf.h>
./modules/hal/stm32/scripts/tests/genllheaders/data/stm32_ll_usart.h:#include <autoconf.h>
./modules/hal/stm32/scripts/genllheaders/header-template.j2:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_fmpi2c.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_lptim.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_wwdg.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_icache.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_gpio.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_pwr.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_adc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_bdma.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_spi.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_utils.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_delayblock.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dlyb.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_lpuart.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_opamp.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_ipcc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_hsem.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_rcc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dmamux.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dac.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_i3c.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dma2d.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_i2c.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_system.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_comp.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_hrtim.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_swpmi.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_tim.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_crc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_cortex.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_ucpd.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_cordic.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dcache.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_sdmmc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_lpgpio.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_fmac.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_usart.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_exti.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_rng.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_mdma.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_iwdg.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_dma.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_bus.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_crs.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_fmc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_pka.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_rtc.h:#include <autoconf.h>
./modules/hal/stm32/stm32cube/common_ll/include/stm32_ll_fsmc.h:#include <autoconf.h>
./modules/hal/nxp/s32/mcux/devices/S32K344/S32K344_features.h:#include "autoconf.h"
```

ycsin commented:
Seems like it works this time, not sure what I missed last time for the `autoconf.h` changes. cc @carlescufi

ycsin commented:
rebased to fix conflicts

ycsin commented:
The PR branchs in `percepio` are now merged while this branch isn't, which is likely to cause havoc to percepio users?

nordicjm commented:
Needs rebase

cfriedt reviewed (dismissed):
This needs a rebase. It would be good to get in prior to v3.7.0

ycsin commented:
Removed `Architecture Review` as it had been reviewed a few months ago, it just requires approvals from the maintainers now.

henrikbrixandersen commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -22,6 +22,36 @@ Build System
+     :header-rows: 1
+
+     * - Affected header files
+     * - ``app_version``
+     * - ``autoconf``
+     * - ``cmake_intdef``
+     * - ``core-isa-dM``
+     * - ``devicetree_generated``
+     * - ``driver-validation``
+     * - ``kobj-types-enum``
+     * - ``linker-kobject-prebuilt-data``
+     * - ``mcuboot_version``
+     * - ``offsets``
+     * - ``otype-to-size``
+     * - ``otype-to-str``
+     * - ``strerror_table``
+     * - ``strsignal_table``
+     * - ``syscall_list``
+     * - ``version``
+     * - ``zsr``
```
Should these have a `.h` postfix?

ycsin replied:
The omission is actually intentional (implied by them being header files) but I can add them if you think it increases clarity

ycsin replied:
Fixed

henrikbrixandersen commented on Kconfig.zephyr:
```diff
@@ -1048,3 +1048,18 @@ config BOOTLOADER_BOSSA_ADAFRUIT_UF2
+
+config LEGACY_GENERATED_INCLUDE_PATH
+	bool "Legacy include path for generated headers"
+	default y
```
Shouldn't this be disabled by default? Otherwise we risk introducing new include statements using the old path.

ycsin replied:
Removed `default y` and added `select DEPRECATED`

ycsin commented:
Rebased and ran the script to update some new #include additions

henrikbrixandersen commented:
@tejlmand Please revisit.

ycsin commented:
Things are failing now because `LEGACY_GENERATED_INCLUDE_PATH` is now disabled by default (previously it was enabled by default).

Currently I'm just trying to see if `hal_stm32` is the only one affected by the latest change, I think it probably makes sense to enable the `LEGACY_GENERATED_INCLUDE_PATH` by default so that this PR has less dependencies on the external modules, and disable + deprecate that after we have all PRs in external modules repo merged.

ycsin commented:
- [ ] wait for CI to pass (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=n`)
- [ ] Remove last commit (4b03a50bcd49ca4bffc2accfda7283b04f317371) (
- [ ] Wait for CI to pass again (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=y`)

henrikbrixandersen commented:
> - [ ] wait for CI to pass (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=n`)
> 
> - [ ] Remove last commit (4b03a50bcd49ca4bffc2accfda7283b04f317371)
> 
> - [ ] Wait for CI to pass again (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=y`)

If CI passes with `CONFIG_LEGACY_GENERATED_INCLUDE_PATH=n`, why revert it?

ycsin commented:
> > * [ ]  wait for CI to pass (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=n`)
> > * [ ]  Remove last commit ([4b03a50](https://github.com/zephyrproject-rtos/zephyr/commit/4b03a50bcd49ca4bffc2accfda7283b04f317371))
> > * [ ]  Wait for CI to pass again (`CONFIG_LEGACY_GENERATED_INCLUDE_PATH=y`)
> 
> If CI passes with `CONFIG_LEGACY_GENERATED_INCLUDE_PATH=n`, why revert it?

The CI pass because the last commit updates the modules to point at my PRs. Correct me if I'm wrong, the `west.yml` will only be updated when these PRs are merged, so in the mean time we need to enable `CONFIG_LEGACY_GENERATED_INCLUDE_PATH` until they are merged and `west.yml` updated.

I also agree with @gmarull that the downstream shouldn't break immediately when they sync to the latest `main` branch after this is merge, i.e. a nightly CI build that tracks the latest `main`. We will show them a nice warning message that asks them to update (while we wait for the `west.yml` to be updated anyway):

![Screenshot 2024-05-24 at 2 55 39 PM](https://github.com/zephyrproject-rtos/zephyr/assets/1563974/2f3ccfef-a0b6-4ca8-9043-135e3bd4709a)

Once all upstream stuff have been updated to work properly, we can stop enabling the `CONFIG_LEGACY_GENERATED_INCLUDE_PATH` by default and the downstream app will fail to compile, if they are still using the legacy path. They will have to enable the Kconfig explicitly and be presented with an additional deprecation warning.

ycsin committed:
build: namespace syscall sources to `zephyr/`

Namespace the `syscall_dispatch.c` & `syscall_export_llext.c`
to `zephyr/` as well

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
CI finally passed 🥳
Will revert the last commit now

marc-hb commented on CMakeLists.txt:
```diff
@@ -109,6 +109,15 @@ add_library(zephyr_interface INTERFACE)
 
+if(CONFIG_LEGACY_GENERATED_INCLUDE_PATH)
+  zephyr_include_directories(${PROJECT_BINARY_DIR}/include/generated/zephyr)
+  message(WARNING "
+    Warning: CONFIG_LEGACY_GENERATED_INCLUDE_PATH is currently enabled by default
+    so that user applications can continue to use the legacy include paths for the
+    generated headers. This Kconfig will be deprecated and eventually removed in
+    the future releases.")
```
Wait... there are _other_ people who look at build warnings? @henrikbrixandersen, you and I should start a club!

I've seen Kconfig and Device tree warnings pretty much every time I build anything more complex than samples/hello_world/... neither of these tools seems to have any `-Werror`, which in my experience is the only way to make anyone care.

marc-hb replied:
The warning will be printed only for people who should fix their code ASAP so printing it every time looks good to me.

ycsin replied:
Otherwise the `CONFIG_LEGACY_GENERATED_INCLUDE_PATH` which is enabled by default now will just make everything passes silently. Most people might not notice that the headers are now relocated to `zephyr/` unless they looked at their build directory or the migration guide.

On the other extreme, if we disable `CONFIG_LEGACY_GENERATED_INCLUDE_PATH` by default, the downstream apps will fail to build immediately, which I don't think is the right way to do in an OS of this scale.

The current implementation IMHO is the middle approach - we don't want to break downstream applications, but at the same time we want to make sure that developers are made aware of the changes, and given enough time to do their migration (as we do ours - some of the PRs aren't merged, `west.yml` not updated to point at them).

ycsin commented:
@cfriedt @nordicjm @finikorg @fabiobaltieri @de-nordic @dcpleung @mbolivar-ampere 
This PR is now ready (again), could you please take another look again?

@gmarull this PR should align with #41543 and consistent with the header migration PR that you have made in the past (~v3.1.0?), could you please review?

de-nordic reviewed (approved):
Changes to nRF QSPI NOR look ok.

fabiobaltieri commented:
This may use some coordination help from someone from @zephyrproject-rtos/release, plenty of dependencies, plus probably something will pop up after merge as well, worth keeping it in mind.

ycsin commented:
ping assignee @tejlmand, PTAL, thanks!

ycsin commented:
> ping assignee @tejlmand, PTAL, thanks!

another ping @tejlmand

ycsin commented:
🥳🍀🤞

marc-hb commented:
Just FYI https://github.com/thesofproject/sof/actions/runs/9278328176/job/25529164408

```
FileNotFoundError:  No such file or directory:
  '/zep_workspace/build-lnl/zephyr/include/generated/autoconf.h'
```

That's because one Python script in SOF extracts values from `autoconf.h` after the build.

The C compilation was fine and the CONFIG_LEGACY_GENERATED_INCLUDE_PATH warning was printed by CMake as expected.

Just FYI.
